File: realm/realm-library/src/main/java/io/realm/coroutines/RealmFlowFactory.java
/*
 * Copyright 2020 Realm Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.realm.coroutines;

import javax.annotation.Nonnull;

import io.realm.DynamicRealm;
import io.realm.DynamicRealmObject;
import io.realm.Realm;
import io.realm.RealmList;
import io.realm.RealmModel;
import io.realm.RealmResults;
import io.realm.annotations.Beta;
import io.realm.internal.coroutines.InternalFlowFactory;
import io.realm.rx.CollectionChange;
import io.realm.rx.ObjectChange;
import kotlinx.coroutines.flow.Flow;

/**
 * Factory class used to create coroutine {@link Flow}s.
 * <p>
 * This class is used by default unless overridden in {@link io.realm.RealmConfiguration.Builder#flowFactory(FlowFactory)}.
 */
@Beta
public class RealmFlowFactory implements FlowFactory {

    private final InternalFlowFactory factory;

    /**
     * Constructor for the Flow factory.
     *
     * @param returnFrozenObjects whether the emissions should return frozen objects or not.
     */
    public RealmFlowFactory(Boolean returnFrozenObjects) {
        this.factory = new InternalFlowFactory(returnFrozenObjects);
    }

    @Override
    public Flow<Realm> from(@Nonnull Realm realm) {
        return factory.from(realm);
    }

    @Override
    public Flow<DynamicRealm> from(@Nonnull DynamicRealm dynamicRealm) {
        return factory.from(dynamicRealm);
    }

    @Override
    public <T> Flow<RealmResults<T>> from(@Nonnull Realm realm, @Nonnull RealmResults<T> results) {
        return factory.from(realm, results);
    }

    @Override
    public <T> Flow<CollectionChange<RealmResults<T>>> changesetFrom(@Nonnull Realm realm, @Nonnull RealmResults<T> results) {
        return factory.changesetFrom(realm, results);
    }

    @Override
    public <T> Flow<RealmResults<T>> from(@Nonnull DynamicRealm dynamicRealm, @Nonnull RealmResults<T> results) {
        return factory.from(dynamicRealm, results);
    }

    @Override
    public <T> Flow<CollectionChange<RealmResults<T>>> changesetFrom(@Nonnull DynamicRealm dynamicRealm, @Nonnull RealmResults<T> results) {
        return factory.changesetFrom(dynamicRealm, results);
    }

    @Override
    public <T> Flow<RealmList<T>> from(@Nonnull Realm realm, @Nonnull RealmList<T> realmList) {
        return factory.from(realm, realmList);
    }

    @Override
    public <T> Flow<CollectionChange<RealmList<T>>> changesetFrom(@Nonnull Realm realm, @Nonnull RealmList<T> list) {
        return factory.changesetFrom(realm, list);
    }

    @Override
    public <T> Flow<RealmList<T>> from(@Nonnull DynamicRealm dynamicRealm, @Nonnull RealmList<T> realmList) {
        return factory.from(dynamicRealm, realmList);
    }

    @Override
    public <T> Flow<CollectionChange<RealmList<T>>> changesetFrom(@Nonnull DynamicRealm dynamicRealm, @Nonnull RealmList<T> list) {
        return factory.changesetFrom(dynamicRealm, list);
    }

    @Override
    public <T extends RealmModel> Flow<T> from(@Nonnull Realm realm, @Nonnull T realmObject) {
        return factory.from(realm, realmObject);
    }

    @Override
    public <T extends RealmModel> Flow<ObjectChange<T>> changesetFrom(@Nonnull Realm realm, @Nonnull T realmObject) {
        return factory.changesetFrom(realm, realmObject);
    }

    @Override
    public Flow<DynamicRealmObject> from(@Nonnull DynamicRealm dynamicRealm, @Nonnull DynamicRealmObject dynamicRealmObject) {
        return factory.from(dynamicRealm, dynamicRealmObject);
    }

    @Override
    public Flow<ObjectChange<DynamicRealmObject>> changesetFrom(@Nonnull DynamicRealm dynamicRealm, @Nonnull DynamicRealmObject dynamicRealmObject) {
        return factory.changesetFrom(dynamicRealm, dynamicRealmObject);
    }
}

File: realm/realm-library/src/main/java/io/realm/coroutines/RealmFlowFactory.kt
/*
 * Copyright 2020 Realm Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.realm.coroutines

import io.realm.*
import io.realm.annotations.Beta
import kotlinx.coroutines.channels.awaitClose
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.callbackFlow
import kotlinx.coroutines.flow.flowOf

/**
 * Factory class used to create coroutine [Flow]s.
 *
 * This class is used by default unless overridden in [RealmConfiguration.Builder.flowFactory].
 */
@Beta
class RealmFlowFactory(
        private val returnFrozenObjects: Boolean = true
) : FlowFactory {

    @Beta
    override fun from(realm: Realm): Flow<Realm> {
        if (realm.isFrozen) {
            return flowOf(realm)
        }

        return callbackFlow {
            val flowRealm = Realm.getInstance(realm.configuration)
            val listener = RealmChangeListener<Realm> { listenerRealm ->
                if (returnFrozenObjects) {
                    offer(realm.freeze())
                } else {
                    offer(listenerRealm)
                }
            }

            flowRealm.addChangeListener(listener)

            if (returnFrozenObjects) {
                offer(flowRealm.freeze())
            } else {
                offer(flowRealm)
            }

            awaitClose {
                flowRealm.removeChangeListener(listener)
                flowRealm.close()
            }
        }
    }

    @Beta
    override fun from(dynamicRealm: DynamicRealm): Flow<DynamicRealm> {
        if (dynamicRealm.isFrozen) {
            return flowOf(dynamicRealm)
        }

        return callbackFlow {
            val flowRealm = DynamicRealm.getInstance(dynamicRealm.configuration)
            val listener = RealmChangeListener<DynamicRealm> { listenerRealm ->
                if (returnFrozenObjects) {
                    offer(dynamicRealm.freeze())
                } else {
                    offer(listenerRealm)
                }
            }

            flowRealm.addChangeListener(listener)

            if (returnFrozenObjects) {
                offer(flowRealm.freeze())
            } else {
                offer(flowRealm)
            }

            awaitClose {
                flowRealm.removeChangeListener(listener)
                flowRealm.close()
            }
        }
    }

    @Beta
    override fun <T> from(realm: Realm, results: RealmResults<T>): Flow<RealmResults<T>> {
        // Return "as is" if frozen, there will be no listening for changes
        if (realm.isFrozen) {
            return flowOf(results)
        }

        val config = realm.configuration

        return callbackFlow {
            // Do nothing if the results are invalid
            if (!results.isValid) {
                return@callbackFlow
            }

            // Get instance to ensure the Realm is open for as long as we are listening
            val flowRealm = Realm.getInstance(config)
            val listener = RealmChangeListener<RealmResults<T>> { listenerResults ->
                if (returnFrozenObjects) {
                    offer(listenerResults.freeze())
                } else {
                    offer(listenerResults)
                }
            }

            results.addChangeListener(listener)

            // Emit current value
            if (returnFrozenObjects) {
                offer(results.freeze())
            } else {
                offer(results)
            }

            awaitClose {
                // Remove listener and cleanup
                if (!flowRealm.isClosed) {
                    results.removeChangeListener(listener)
                    flowRealm.close()
                }
            }
        }
    }

    override fun <T> from(dynamicRealm: DynamicRealm, results: RealmResults<T>): Flow<RealmResults<T>> {
        // Return "as is" if frozen, there will be no listening for changes
        if (dynamicRealm.isFrozen) {
            return flowOf(results)
        }

        val config = dynamicRealm.configuration

        return callbackFlow {
            // Do nothing if the results are invalid
            if (!results.isValid) {
                return@callbackFlow
            }

            // Get instance to ensure the Realm is open for as long as we are listening
            val flowRealm = DynamicRealm.getInstance(config)
            val listener = RealmChangeListener<RealmResults<T>> { listenerResults ->
                if (returnFrozenObjects) {
                    offer(listenerResults.freeze())
                } else {
                    offer(listenerResults)
                }
            }

            results.addChangeListener(listener)

            // Emit current value
            if (returnFrozenObjects) {
                offer(results.freeze())
            } else {
                offer(results)
            }

            awaitClose {
                // Remove listener and cleanup
                if (!flowRealm.isClosed) {
                    results.removeChangeListener(listener)
                    flowRealm.close()
                }
            }
        }
    }

    @Beta
    override fun <T> from(realm: Realm, realmList: RealmList<T>): Flow<RealmList<T>> {
        // Return "as is" if frozen, there will be no listening for changes
        if (realm.isFrozen) {
            return flowOf(realmList)
        }

        val config = realm.configuration

        return callbackFlow {
            // Do nothing if the results are invalid
            if (!realmList.isValid) {
                return@callbackFlow
            }

            // Get instance to ensure the Realm is open for as long as we are listening
            val flowRealm = Realm.getInstance(config)
            val listener = RealmChangeListener<RealmList<T>> { listenerResults ->
                if (returnFrozenObjects) {
                    offer(listenerResults.freeze())
                } else {
                    offer(listenerResults)
                }
            }

            realmList.addChangeListener(listener)

            // Emit current value
            if (returnFrozenObjects) {
                offer(realmList.freeze())
            } else {
                offer(realmList)
            }

            awaitClose {
                // Remove listener and cleanup
                if (!flowRealm.isClosed) {
                    realmList.removeChangeListener(listener)
                    flowRealm.close()
                }
            }
        }
    }

    override fun <T> from(dynamicRealm: DynamicRealm, realmList: RealmList<T>): Flow<RealmList<T>> {
        // Return "as is" if frozen, there will be no listening for changes
        if (dynamicRealm.isFrozen) {
            return flowOf(realmList)
        }

        val config = dynamicRealm.configuration

        return callbackFlow {
            // Do nothing if the results are invalid
            if (!realmList.isValid) {
                return@callbackFlow
            }

            // Get instance to ensure the Realm is open for as long as we are listening
            val flowRealm = DynamicRealm.getInstance(config)
            val listener = RealmChangeListener<RealmList<T>> { listenerResults ->
                if (returnFrozenObjects) {
                    offer(listenerResults.freeze())
                } else {
                    offer(listenerResults)
                }
            }

            realmList.addChangeListener(listener)

            // Emit current value
            if (returnFrozenObjects) {
                offer(realmList.freeze())
            } else {
                offer(realmList)
            }

            awaitClose {
                // Remove listener and cleanup
                if (!flowRealm.isClosed) {
                    realmList.removeChangeListener(listener)
                    flowRealm.close()
                }
            }
        }
    }

    @Beta
    override fun <T : RealmModel> from(realm: Realm, realmObject: T): Flow<T> {
        // Return "as is" if frozen, there will be no listening for changes
        if (realm.isFrozen) {
            return flowOf(realmObject)
        }

        val config = realm.configuration

        return callbackFlow<T> {
            // Do nothing if the object is invalid
            if (!RealmObject.isValid(realmObject)) {
                return@callbackFlow
            }

            // Get instance to ensure the Realm is open for as long as we are listening
            val flowRealm = Realm.getInstance(config)
            val listener = RealmChangeListener<T> { listenerObj ->
                if (returnFrozenObjects) {
                    offer(RealmObject.freeze(listenerObj) as T)
                } else {
                    offer(listenerObj)
                }
            }

            RealmObject.addChangeListener(realmObject, listener)

            // Emit current value
            if (returnFrozenObjects) {
                offer(RealmObject.freeze(realmObject))
            } else {
                offer(realmObject)
            }

            awaitClose {
                // Remove listener and cleanup
                if (!flowRealm.isClosed) {
                    RealmObject.removeChangeListener(realmObject, listener)
                    flowRealm.close()
                }
            }
        }
    }

    @Beta
    override fun from(dynamicRealm: DynamicRealm, dynamicRealmObject: DynamicRealmObject): Flow<DynamicRealmObject> {
        // Return "as is" if frozen, there will be no listening for changes
        if (dynamicRealm.isFrozen) {
            return flowOf(dynamicRealmObject)
        }

        val config = dynamicRealm.configuration

        return callbackFlow<DynamicRealmObject> {
            // Do nothing if the object is invalid
            if (!dynamicRealmObject.isValid) {
                return@callbackFlow
            }

            // Get instance to ensure the Realm is open for as long as we are listening
            val flowRealm = DynamicRealm.getInstance(config)
            val listener = RealmChangeListener<DynamicRealmObject> { listenerObj ->
                if (returnFrozenObjects) {
                    offer(listenerObj.freeze())
                } else {
                    offer(listenerObj)
                }
            }

            dynamicRealmObject.addChangeListener(listener)

            // Emit current value
            if (returnFrozenObjects) {
                offer(RealmObject.freeze(dynamicRealmObject))
            } else {
                offer(dynamicRealmObject)
            }

            awaitClose {
                // Remove listener and cleanup
                if (!flowRealm.isClosed) {
                    dynamicRealmObject.removeChangeListener(listener)
                    flowRealm.close()
                }
            }
        }
    }
}
